sql/src/main/java/io/crate/planner/operators/Union.java

```diff
@@ -0,0 +1,211 @@
+
+            LogicalPlan lhsPlan = LogicalPlanner.plan(left, FetchMode.NEVER_CLEAR, false).build(tableStats, usedFromLeft);
+            LogicalPlan rhsPlan = LogicalPlanner.plan(right, FetchMode.NEVER_CLEAR, false).build(tableStats, usedFromRight);
+
+            return new Union(lhsPlan, rhsPlan);
+        };
+    }
+
+    private static void addColumnsFrom(Iterable<? extends Symbol> symbols,
+                                       Consumer<? super Symbol> consumer,
+                                       QueriedRelation rel) {
+
+        for (Symbol symbol : symbols) {
+            addColumnsFrom(symbol, consumer, rel);
+        }
+    }
+
+    private static void addColumnsFrom(@Nullable Symbol symbol, Consumer<? super Symbol> consumer, QueriedRelation rel) {
+        if (symbol == null) {
+            return;
+        }
+        FieldsVisitor.visitFields(symbol, f -> {
+            if (f.relation().getQualifiedName().equals(rel.getQualifiedName())) {
+                consumer.accept(rel.querySpec().outputs().get(f.index()));
+            }
+        });
+    }
+
+    private Union(LogicalPlan lhs, LogicalPlan rhs) {
+        this.lhs = lhs;
+        this.rhs = rhs;
+        this.baseTables = new ArrayList<>();
+        this.baseTables.addAll(lhs.baseTables());
+        this.baseTables.addAll(rhs.baseTables());
+        this.expressionMapping = new HashMap<>();
+        this.expressionMapping.putAll(lhs.expressionMapping());
+        this.expressionMapping.putAll(rhs.expressionMapping());
+    }
+
+    @Override
+    public Plan build(Planner.Context plannerContext,
+                      ProjectionBuilder projectionBuilder,
+                      int limit,
+                      int offset,
+                      @Nullable OrderBy order,
+                      @Nullable Integer pageSizeHint) {
+
+        Integer childPageSizeHint = limit != TopN.NO_LIMIT
+            ? limitAndOffset(limit, offset)
+            : null;
+
+        Plan left = lhs.build(plannerContext, projectionBuilder, limit + offset, offset, null, childPageSizeHint);
+        Plan right = rhs.build(plannerContext, projectionBuilder, limit + offset, offset, null, childPageSizeHint);
+
+        left = addMergeIfNeeded(left, plannerContext);
+        right = addMergeIfNeeded(right, plannerContext);
+
+        ResultDescription leftResultDesc = left.resultDescription();
+        ResultDescription rightResultDesc = right.resultDescription();
+
+        MergePhase mergePhase = new MergePhase(
+            plannerContext.jobId(),
+            plannerContext.nextExecutionPhaseId(),
+            "union",
+            leftResultDesc.nodeIds().size() + rightResultDesc.nodeIds().size(),
+            2,
+            Collections.singletonList(plannerContext.handlerNode()),
+            leftResultDesc.streamOutputs(),
+            Collections.emptyList(),
+            DistributionInfo.DEFAULT_BROADCAST,
+            null
+        );
+
+
+        return new UnionPlan(
```

Comment:
Hm - what we haven't considered here is that if we do a insert from subquery, we're unnecessarily moving all data to the handler before creating the insert requests - it would be possible to keep the data distributed and insert from each node directly (as long as there is no limit involved)

I guess this is okay for the initial support - but we'll have to consider how we can adapt the LogicalPlan building to support such cases. One option would be to make the `ensureOnHandler` logic part of the `LogicalPlan` - so that we can encapsulate different behaviours depending on the operator.

Reply:
>Hm - what we haven't considered here is that if we do a insert from subquery, we're unnecessarily moving all data to the handler before creating the insert requests 

This is a general limitation of insert from subqueries because of https://github.com/mxm/crate/blob/maser/sql/src/main/java/io/crate/planner/consumer/InsertFromSubQueryConsumer.java#L92 , right?

```java
Insert into t values (id)
select id from t2
union all
select id from t3
```

>One option would be to make the ensureOnHandler logic part of the LogicalPlan - so that we can encapsulate different behaviours depending on the operator.

Seems like the best way to let the plan decide whether it needs a final merge.

Reply:
> This is a general limitation of insert from subqueries because of https://github.com/mxm/crate/blob/maser/sql/src/main/java/io/crate/planner/consumer/InsertFromSubQueryConsumer.java#L92 , right?

The `ensureOnHandler` call happens after the projection has been added. So no, this is not a general limitation.


> Seems like the best way to let the plan decide whether it needs a final merge.

So far we always tried to keep the executionPlan returned on `build`
distributed if possible. The *parent* operators were always responsible to add
the merge if needed. The problem is that with union we currently can't
represent it in a distribution way and let the parents decide if a merge is
required. This will need some interface extension / remodelling.

But as I said - I think this is out of scope of the work here. But we've to think about how we do this later on.

Reply:
I think this whole logic of when a merge is applied or not, probably needs to be reworked in the future. As you said, the parents have to check but it would be nice if the children already decided if a merge is necessary for the following operations.

>The problem is that with union we currently can't
represent it in a distribution way and let the parents decide if a merge is
required.

I think I don't understand. The `Union` currently decides if a merge is required.